src/workerd/util/string-buffer.h

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2023 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+#pragma once
+
+#include <cstring>
+#include <kj/string.h>
+#include <kj/vector.h>
+
+namespace workerd {
+
+// String buffer optimized for appending a lot of strings together.
+// Allocates StackSize chunk on the stack and uses that until full.
+// Keeps allocating new chunks of at least HeapChunkSize as needed.
+// Doesn't perform any heap allocations if string stays within
+// StackSize bytes (without \0)
+template<size_t StackSize>
+class StringBuffer {
+
+public:
+  KJ_DISALLOW_COPY_AND_MOVE(StringBuffer);
+
+  explicit StringBuffer(size_t heapChunkSize): heapChunkSize(heapChunkSize), tail(&arr[0]), cap(StackSize) {}
+
+  void append() {}
+
+  template <typename First, typename... Rest>
+  void append(First&& first, Rest&&...rest) {
+    appendImpl(kj::fwd<First>(first));
+    append(kj::fwd<Rest>(rest)...);
+  }
+
+  kj::String toString() {
+    auto result = kj::heapString(len);
+    copyTo(result.begin());
+    return result;
+  }
+
+private:
+  // minimum heap chunk size
+  const size_t heapChunkSize;
+
+  // chunk on the stack
+  char arr[StackSize];
+
+  // on the heap chunks
+  kj::Vector<kj::Array<char>> chunks;
```

Comment:
Non-blocking... and just thinking out loud....  In Node.js we have a utility type called `MaybeStackBuffer<T>` that encapsulates this Maybe-Stack-Maybe-Heap pattern. It's a lot like `KJ_STACK_ARRAY` but in a form that is easy to define and use as class members like this. It might be worthwhile to adopt a version of that for our use.